src/promotion/nhit/nhit.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright(c) 2019 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+#include "hash.h"
+#include "../../metadata/metadata.h"
+#include "../../ocf_priv.h"
+#include "../../engine/engine_common.h"
+
+#include "nhit.h"
+#include "../ops.h"
+
+#define NHIT_MAPPING_RATIO 2
+
+#define NHIT_MIN_THRESHOLD 1
+#define NHIT_MAX_THRESHOLD 1000
+#define NHIT_THRESHOLD_DEFAULT 3
+
+struct nhit_policy_context {
+	nhit_hash_t hash_map;
+
+	/* Configurable parameters */
+	env_atomic insertion_threshold;
+};
+
+ocf_error_t nhit_init(ocf_cache_t cache, ocf_promotion_policy_t policy)
+{
+	struct nhit_policy_context *ctx;
+	int result = 0;
+
+	ctx = env_vmalloc(sizeof(*ctx));
+	if (!ctx) {
+		result = -OCF_ERR_NO_MEM;
+		goto exit;
+	}
+	policy->ctx = ctx;
```

Comment:
How about moving this assignment to line 45? Context is fully initialized at that point.

Reply:
True, fixed